sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mlme.c

```diff
@@ -0,0 +1,354 @@
+        default:
+            mlme_confirm->status = -EINVAL;
+            break;
+    }
+}
+
+void gnrc_lorawan_mlme_request(gnrc_lorawan_t *mac, mlme_request_t *mlme_request,
+                               mlme_confirm_t *mlme_confirm)
+{
+    switch (mlme_request->type) {
+        case MLME_JOIN:
+            if(mac->mlme.activation != MLME_ACTIVATION_NONE) {
+                mlme_confirm->status = -EINVAL;
+                return;
+            }
+            if (!gnrc_lorawan_mac_acquire(mac)) {
+                mlme_confirm->status = -EBUSY;
+                return;
+            }
+
+            if (mac->mlme.backoff_budget < 0) {
+                mlme_confirm->status = -EDQUOT;
+                break;
+            }
+            memcpy(mac->appskey, mlme_request->param.join.appkey, LORAMAC_APPKEY_LEN);
+            mlme_confirm->status = gnrc_lorawan_send_join_request(mac, mlme_request->param.join.deveui,
+                                                                  mlme_request->param.join.appeui, mlme_request->param.join.appkey, mlme_request->param.join.dr);
+            break;
+        case MLME_LINK_CHECK:
+            mac->mlme.pending_mlme_opts |= GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_DEFERRED;
+            break;
+        case MLME_SET:
+            _mlme_set(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_GET:
+            _mlme_get(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_RESET:
+            gnrc_lorawan_reset(mac);
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+            break;
+        default:
+            break;
+    }
+}
+
+int _fopts_mlme_link_check_req(gnrc_lorawan_t *mac, lorawan_buffer_t *buf)
+{
+    (void) mac;
+    if (buf) {
+        assert(buf->index + GNRC_LORAWAN_CID_SIZE <= buf->size);
+        buf->data[buf->index++] = GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS;
+    }
+
+    return GNRC_LORAWAN_CID_SIZE;
+}
+
+static int _mlme_link_check_ans(gnrc_lorawan_t *mac, lorawan_buffer_t *fopt)
+{
+    if (fopt->index + GNRC_LORAWAN_FOPT_LINK_ANS_SIZE > fopt->size) {
+        return -EINVAL;
+    }
+    fopt->index++;
+
+    mlme_confirm_t *mlme_confirm = _mlme_allocate(mac);
+    mlme_confirm->param.link_req.margin = fopt->data[fopt->index++];
+    mlme_confirm->param.link_req.num_gateways = fopt->data[fopt->index++];
+
+    mlme_confirm->type = MLME_LINK_CHECK;
+    mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+    mac->netdev.event_callback(&mac->netdev, NETDEV_EVENT_MLME_CONFIRM);
+
+    mac->mlme.pending_mlme_opts &= ~GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+
+    return 0;
+}
+
+void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size)
+{
+    if (!fopts || !size) {
+        goto out;
+    }
+
+    lorawan_buffer_t buf;
+    _buffer_reset(&buf, fopts, size);
+
+    while (buf.index < buf.size) {
+        switch (*(buf.data)) {
```

Comment:
switch looks like overkill here, `if ... else` looks cleaner (and avoids `goto`):
```
while (buf.index < buf.size) {
    if ((*(buf.data) != GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS) ||
        (_mlme_link_check_ans(mac, &buf) < 0)) {
        break;
    }
}
```

Reply:
I believe the code above should do the same as your `switch` + `goto`🤞

Reply:
This list is intended to be expanded, so I will need a switch anyway :(. I removed the goto in the last fixup though